editor/test/integration/save_test.clj

```diff
@@ -150,23 +135,46 @@
-                    pb-disk (protobuf/read-text pb-class resource)
-                    path []
-                    [disk-diff save-diff both] (data/diff (get-in pb-disk path) (get-in pb-save path))]
-                (is (nil? disk-diff))
-                (is (nil? save-diff))
-                (when (and (nil? disk-diff) (nil? save-diff))
-                  (let [diff-lines (keep (fn [[f s]] (when (not= f s) [f s])) (map vector (str/split-lines file) (str/split-lines (:content save))))]
-                    (doseq [[f s] diff-lines]
-                      (prn "f" f)
-                      (prn "s" s))))))
-            (is (= file (:content save)))))))))
+                save-data (g/node-value node-id :save-data)
+                save-string (:content save-data)
+                disk-string (slurp resource)
+                print-line-diff! (fn print-line-diff! []
+                                   (let [diff-lines
+                                         (keep (fn [[f s]]
+                                                 (when (not= f s)
+                                                   [f s]))
+                                               (map vector
+                                                    (str/split-lines disk-string)
+                                                    (str/split-lines save-string)))]
+                                     (println "When comparing" path)
+                                     (doseq [[disk-line save-line] diff-lines]
+                                       (prn 'disk disk-line)
+                                       (prn 'save save-line))))]
+            (is (not (g/error? save-data)))
+            (when-not (is (= disk-string save-string))
+              (if-some [read-raw-fn (:read-raw-fn (resource/resource-type resource))]
+
+                ;; We have a read-fn. Compare the read data representations.
+                (let [disk-pb-msg (read-raw-fn resource)
+                      save-pb-msg (with-open [reader (StringReader. save-string)]
+                                    (read-raw-fn reader))
+                      [only-in-disk-pb-msg only-in-save-pb-msg] (data/diff disk-pb-msg save-pb-msg)]
+                  (is (nil? only-in-disk-pb-msg))
+                  (is (nil? only-in-save-pb-msg))
+                  (if (and (nil? only-in-disk-pb-msg)
+                           (nil? only-in-save-pb-msg))
+
+                    ;; The strings differ, but the read data does not. Print a line diff.
+                    (print-line-diff!)
+
+                    ;; The read data differs. Print a data diff.
+                    (do
+                      (println "When comparing" path)
+                      (prn 'disk only-in-disk-pb-msg)
+                      (prn 'save only-in-save-pb-msg))))
+
+                ;; We don't have a read-fn. Print a line diff.
+                (print-line-diff!)))))))))
```

Comment:
This test should have the same behavior as before, but was rewritten to provide more output about what differs.